Sources/Fuzzilli/Corpus/MarkovCorpus.swift

```diff
@@ -0,0 +1,168 @@
+import Foundation
+
+/// Corpus & Scheduler based on 
+/// Coverage-based Greybox Fuzzing as Markov Chain paper
+/// https://mboehme.github.io/paper/TSE18.pdf
+/// Simply put, the corpus keeps track of which paths have been found, and prioritizes seeds
+/// whose path has been hit less than average. Ideally, this allows the fuzzer to prioritize
+/// less explored coverage
+
+public class MarkovCorpus: ComponentBase, Corpus {
+
+    private var allIncludedPrograms: [Program] // All programs currently in the corpus
+    private var programExecutionQueue: [Program]  // Queue of programs to be executed next, all of which hit a rare edge
+
+    private var edgeMap: [UInt64:Program]
+
+    private var totalExecs: UInt64
+
+    private var currentProg: Program?
+
+    private var remainingEnergy: UInt64
+
+    public init() {
+        self.allIncludedPrograms = []
+        self.programExecutionQueue = []
+        self.totalExecs = 0
+        self.edgeMap = [:]
+        self.currentProg = nil
+        self.remainingEnergy = 0
+        super.init(name: "MarkovCorpus")
+    }
+    
+    override func initialize() {
+        // The corpus must never be empty. Other components, such as the ProgramBuilder, rely on this
+        if isEmpty {
+            for _ in 1...5 {
+                let b = fuzzer.makeBuilder()
+                let objectConstructor = b.loadBuiltin("Object")
+                b.callFunction(objectConstructor, withArgs: [])
+                add(b.finalize(), ProgramAspects(outcome: .succeeded))
+            }
+        }
+    }
+
+    // /// Adds an individual program to the corpus.
+    // /// This method should only be called for programs from outside sources, such as other connected workers, and imports from disk
+    // /// to ensure that new edges are acquired/tracked properly.
+    // public func add(_ program: Program) {
+    //     guard program.size > 0 else { return }
+    //     let execution = fuzzer.execute(program)
+    //     guard execution.outcome == .succeeded else { return }
+    //     if let aspects = fuzzer.evaluator.evaluate(execution) {
+    //         add(program, aspects)
+    //     }
+    // }
+
+    // /// Adds multiple programs to the corpus.
+    // public func add(_ programs: [Program]) {
+    //     logger.info("Import of \(programs.count) programs")
+    //     for (index, prog) in programs.enumerated() {
+    //         if index % 500 == 0 {
+    //             logger.info("Markov Corpus import at \(index) of \(programs.count)")
+    //         }
+    //         add(prog)
+    //     }
+    // }
+
+    public func add(_ program: Program, _ aspects: ProgramAspects) {
+        guard program.size > 0 else { return }
+        allIncludedPrograms.append(program)
+        let edges = aspects.toEdges()
```

Comment:
I think here (and generally in this file) you can just downcast to a `CovEdgeSet` (and log an error if the downcast fails), and in `initialize()` make sure that the evaluator is a `ProgramCoverageEvaluator` (otherwise terminate through a `logger.fatal`), and maybe keep a reference to it so there are no further downcasts required for it. It seems like the MarkovCorpus can only work properly with the coverage evaluator, and so it's probably better to just make that dependency explicit (through downcasts) rather than trying to hide it by moving methods (`toEdges`) into common interfaces (`Aspects`) even though they only make sense on one specific implementation (`CovEdgeSet`).